src/itwinai/torch/trainer.py

```diff
@@ -745,6 +758,340 @@ def test_step(
+            self.generator, self.optimizerG, **distribute_kwargs)
+
+    def train_epoch(self, epoch: int):
+        self.discriminator.train()
+        self.generator.train()
+        g_train_losses = []
+        d_train_losses = []
+        for batch_idx, (real_images, _) in enumerate(self.train_dataloader):
+            lossG, lossD = self.train_step(
+                real_images, batch_idx)
+            g_train_losses.append(lossG)
+            d_train_losses.append(lossD)
+            self.log(
+                item=lossG,
+                identifier='generator training loss per epoch',
+                kind='metric',
+                step=self.train_glob_step,
+                batch_idx=batch_idx
+            )
+            self.log(
+                item=lossD,
+                identifier='discriminator training loss per epoch',
+                kind='metric',
+                step=self.train_glob_step,
+                batch_idx=batch_idx
+            )
+
+            self.train_glob_step += 1
+        # Aggregate and log losses
+        avg_g_loss = torch.mean(torch.stack(g_train_losses))
+        self.log(
+            item=avg_g_loss.item(),
+            identifier='g_train_loss_av per batch',
+            kind='metric',
+            step=epoch,
+        )
+
+        avg_d_loss = torch.mean(torch.stack(d_train_losses))
+        self.log(
+            item=avg_d_loss.item(),
+            identifier='d_train_loss_av per batch',
+            kind='metric',
+            step=epoch,
+        )
+
+        self.save_plots_and_images(epoch)
+
+    def validation_epoch(self, epoch: int):
+        g_validation_losses = []
+        g_validation_accuracy = []
+        self.discriminator.eval()
+        self.generator.eval()
+        validation_loss = torch.tensor(0.0, device=self.device)
+        for batch_idx, (real_images, _) in enumerate(
+                self.validation_dataloader):
+            loss_gen, accuracy_gen = self.validation_step(
+                real_images, batch_idx)
+            g_validation_losses.append(loss_gen)
+            g_validation_accuracy.append(accuracy_gen)
+            self.validation_glob_step += 1
+
+        # Aggregate and log metrics
+        validation_loss = torch.mean(torch.stack(g_validation_losses))
+        self.log(
+            item=validation_loss.item(),
+            identifier='gen_validation_loss_epoch',
+            kind='metric',
+            step=self.validation_glob_step,
+        )
+        validation_accuracy = torch.mean(torch.stack(g_validation_accuracy))
+        self.log(
+            item=validation_accuracy.item(),
+            identifier='gen_validation_accuracy_epoch',
+            kind='metric',
+            step=self.validation_glob_step,
+        )
+
+        return validation_loss
+
+    def train_step(self, real_images, batch_idx):
+        real_images = real_images.to(self.device)
+        batch_size = real_images.size(0)
```

Comment:
I have read more as to why `real_images.size(0)` is a more efficient way to get the batch size compared to getting it statically from the config. This is because it dynamically computes the size of the image batch which might vary on the last batch since the total number of data points might not always be equally divided in all the batches so the last batch could have less images hence the error that was being raised when I use the static batch_size from the config.

Reply:
Good point! You're right. Let's leave it as is. Thanks for checking